File: Problem508/Solution.java
import java.util.*;

public class Solution {
	private int maxCount = Integer.MIN_VALUE;

	public int[] findFrequentTreeSum(TreeNode root) {
		Map<Integer, Integer> map = new HashMap<Integer, Integer>();
		getSums(root, map);
		List<Integer> list = new ArrayList<>();
		map.forEach((k, v) -> {
			if (v == maxCount) {
				list.add(k);
			}
		});
		int[] result = new int[list.size()];
		for (int i = 0; i < list.size(); i++)
			result[i] = list.get(i);
		return result;
	}

	private int getSums(TreeNode root, Map<Integer, Integer> map) {
		int sum = root == null ? 0 : root.val + getSums(root.left, map) + getSums(root.right, map);
		if (root != null) {
			map.put(sum, map.getOrDefault(sum, 0) + 1);
			maxCount = Math.max(maxCount, map.get(sum));
		}
		return sum;
	}
}

File: Problem508/Solution.go
package main

type TreeNode struct {
	Val   int
	Left  TreeNode
	Right TreeNode
}

var max int

func findFrequentTreeSum(root *TreeNode) []int {
	max = 0
	var result []int
	counts := map[int]int{}
	getCounts(root, &counts)
	for k, v := range counts {
		if v == max {
			result = append(result, k)
		}
	}
	return result
}

func getCounts(root *TreeNode, counts *map[int]int) int {
	if root == nil {
		return 0
	}
	var sum int
	var count int
	sum = root.Val + getCounts(root.Left, counts) + getCounts(root.Right, counts)

	count = (*counts)[sum]
	(*counts)[sum] = count + 1
	max = getMax(max, (*counts)[sum])
	return sum
}

func getMax(a int, b int) int {
	if a > b {
		return a
	}
	return b
}
